### PR TITLE
Route task queue user data writes via a single matching node per name…

### PR DESCRIPTION
**Pull request by bergundy**
…space

Required to avoid concurrently writing to the same partition key in Cassandra.

### CODE REVIEW

File: service/matching/taskQueueManager.go [hunk 1]
```diff
@@ -494,27 +491,20 @@ func (c *taskQueueManagerImpl) GetUserData(ctx context.Context) (*persistencespb
...
-		}
+	c.signalIfFatal(err)
+	_, err = c.matchingClient.ReplicateTaskQueueUserData(ctx, &matchingservice.ReplicateTaskQueueUserDataRequest{
+		NamespaceId: c.db.namespaceID.String(),
+		TaskQueue:   c.taskQueueID.BaseNameString(),
+		UserData:    newData.GetData(),
+	})
+	if err != nil {
+		c.logger.Error("Failed to publish a replication task after updating task queue user data", tag.Error(err))
+		return serviceerror.NewUnavailable("storing task queue user data succeeded but publishing to the namespace replication queue failed, please try again")
```
Comment (Sushisource) [inline service/matching/taskQueueManager.go:505]: Not quite direct publishing to queue any more, maybe slight rewording

File: service/matching/matchingEngine.go [hunk 2]
```diff
@@ -115,6 +122,10 @@ type (
 		timeSource           clock.TimeSource
 		// Only set if global namespaces are enabled on the cluster.
 		namespaceReplicationQueue persistence.NamespaceReplicationQueue
+		// Disables concurrent task queue user data updates and replication requests (due to a cassandra limitation)
+		namespaceUpdateLockMap map[string]*namespaceUpdateLocks
+		// Serializes access to the per namespace lock map
+		namespaceUpdateLockMapLock sync.Mutex
```
Comment (Sushisource) [inline service/matching/matchingEngine.go:128]: Not that I have any alternative but man, lockfest 2000

**Review by Sushisource — APPROVED** [review]

File: service/matching/matchingEngine.go (hunk 2 above)
Comment (bergundy) [reply service/matching/matchingEngine.go:128]: Yeah, the future is here

File: service/matching/taskQueueManager.go (hunk 1 above)
Comment (bergundy) [reply service/matching/taskQueueManager.go:505]: This LGTM TBH.

File: service/matching/matchingEngine.go (hunk 2 above)
Comment (dnr) [reply service/matching/matchingEngine.go:128]: `IDMutex` is at least a little cleaner than this

**Pull request merged**